sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteRowLevelCommand.scala

```diff
@@ -68,4 +73,58 @@ trait RewriteRowLevelCommand extends Rule[LogicalPlan] {
       operation.requiredMetadataAttributes,
       relation)
   }
+
+  protected def resolveRowIdAttrs(
+      relation: DataSourceV2Relation,
+      operation: RowLevelOperation): Seq[AttributeReference] = {
+
+    operation match {
+      case supportsDelta: SupportsDelta =>
+        val rowIdAttrs = V2ExpressionUtils.resolveRefs[AttributeReference](
+          supportsDelta.rowId,
+          relation)
+
+        val nullableRowIdAttrs = rowIdAttrs.filter(_.nullable)
+        if (nullableRowIdAttrs.nonEmpty) {
+          throw QueryCompilationErrors.nullableRowIdError(nullableRowIdAttrs)
+        }
+
+        rowIdAttrs
+
+      case _ =>
+        throw QueryCompilationErrors.notDeltaBasedRowLevelOperationError(operation)
```

Comment:
this should never happen, we can throw `SparkException.internalError` directly.

Reply:
Reworked this part and added a safe explicit cast.